eryx/pdb.py

```diff
@@ -316,3 +341,308 @@ def _tile_form_factors(self):
+
+    def _setup_gaussian_network_model(self):
+        """
+        Build interaction pair list and spring constant.
+        """
+        self.build_gamma()
+        self.build_neighbor_list()
+
+    def build_gamma(self):
+        """
+        The spring constant gamma dictates the interaction strength
+        between pairs of atoms. It is defined across the main unit
+        cell and its neighbors, for each asymmetric unit, resulting
+        in an array of shape (n_cell, n_asu, n_asu).
+        The spring constant between atoms belonging to different asus
+        can be set to a different value than that for intra-asus atoms.
+        """
+        self.gamma = np.zeros((self.n_cell, self.n_asu, self.n_asu))
+        for i_asu in range(self.n_asu):
+            for i_cell in range(self.n_cell):
+                for j_asu in range(self.n_asu):
+                    self.gamma[i_cell, i_asu, j_asu] = self.gamma_inter
+                    if (i_cell == self.id_cell_ref) and (j_asu == i_asu):
+                        self.gamma[i_cell, i_asu, j_asu] = self.gamma_intra
+
+    def build_neighbor_list(self):
+        """
+        Returns the list asu_neighbors[i_asu][i_cell][j_asu]
+        of atom pairs between the ASU i_asu from the reference cell
+        and ASU j_asu from the cell i_cell. The length of the returned
+        list is the number of atoms in one ASU.
+        For each atom i in the ASU i_asu in the reference cell,
+        asu_neighbors[i_asu][i_cell][j_asu][i] returns the indices of
+        its neighbors, if any, in the ASU j_asu in cell i_cell.
+        """
+        self.asu_neighbors = []
+
+        for i_asu in range(self.n_asu):
+            self.asu_neighbors.append([])
+            kd_tree1 = KDTree(self.crystal.get_asu_xyz(i_asu, self.crystal.id_to_hkl(self.id_cell_ref)))
+
+            for i_cell in range(self.n_cell):
+                self.asu_neighbors[i_asu].append([])
+
+                for j_asu in range(self.n_asu):
+                    self.asu_neighbors[i_asu][i_cell].append([])
+                    kd_tree2 = KDTree(self.crystal.get_asu_xyz(j_asu, self.crystal.id_to_hkl(i_cell)))
+
+                    self.asu_neighbors[i_asu][i_cell][j_asu] = kd_tree1.query_ball_tree(kd_tree2, r=self.enm_cutoff)
+
+    def compute_hessian(self):
+        """
+        For a pair of atoms the Hessian in a GNM is defined as:
+        1. i not j and dij =< cutoff: -gamma_ij
+        2. i not j and dij > cutoff: 0
+        3. i=j: -sum_{j not i} hessian_ij
+
+        Returns
+        -------
+        hessian: numpy.ndarray,
+                 shape (n_asu, n_atoms_per_asu,
+                        n_cell, n_asu, n_atoms_per_asu)
+                 type 'complex'
+            - dimension 0: index ASUs in reference cell
+            - dimension 1: index their atoms
+            - dimension 2: index neighbor cells
+            - dimension 3: index ASUs in neighbor cell
+            - dimension 4: index atoms in neighbor ASU
```

Comment:
Yikes, I didn't realize the dimensionality of the Hessian was so frighteningly complex! Sounds like a broadcasting nightmare waiting to happen...

Reply:
Yes I've been back and forth with the actual order and ended up with this one - hopefully it won't bite me later!